File: go/set/set.go
package set

import (
	"github.com/stretchr/testify/assert"
	"math/rand/v2"
	"sort"
	"testing"
)

type Set interface {
	Add(val int) bool
	Contains(val int) bool
	Remove(val int) bool
}

func GenericRandomCases(t *testing.T, setTypeName string, newFunc func() Set) {
	var random = rand.New(rand.NewPCG(0, 0))

	for range 10_000 {
		set := newFunc()

		maxVal := random.IntN(100)
		addProb := random.Float64()
		findProb := random.Float64()

		expectedSet := map[int]bool{}

		for range 1_000 {
			val := random.IntN(maxVal + 1)

			if p := random.Float64(); p < addProb {
				expected := !expectedSet[val]
				actual := set.Add(val)
				if actual != expected {
					assert.FailNowf(t, "",
						"%v(%v)\n.Add(%v):\nwant %v\n got %v",
						setTypeName, setToArr(expectedSet), val, expected, actual,
					)
				}
				expectedSet[val] = true
			} else if p < findProb {
				expected := expectedSet[val]
				actual := set.Contains(val)
				if actual != expected {
					assert.FailNowf(t, "",
						"%v(%v)\n.Contains(%v):\nwant %v\n got %v",
						setTypeName, setToArr(expectedSet), val, expected, actual,
					)
				}
			} else {
				expected := expectedSet[val]
				actual := set.Remove(val)
				if actual != expected {
					assert.FailNowf(t, "",
						"%v(%v)\n.Remove(%v):\nwant %v\n got %v",
						setTypeName, setToArr(expectedSet), val, expected, actual,
					)
				}
				delete(expectedSet, val)
			}

		}
	}
}

func setToArr(set map[int]bool) []int {
	var arr []int

	for key := range set {
		arr = append(arr, key)
	}

	sort.Ints(arr)

	return arr
}

File: python/set/set.py
from abc import ABC, abstractmethod
import random as rand


class Set(ABC):
	@abstractmethod
	def add(self, val) -> bool:
		pass

	@abstractmethod
	def contains(self, val) -> bool:
		pass

	@abstractmethod
	def remove(self, val) -> bool:
		pass


def generic_random_cases(type_name, new_func):
	random = rand.Random(0)

	for test in range(10_000):
		actual_set = new_func()

		max_val = random.randrange(100)
		add_prob = random.random()
		find_prob = random.random()

		expected_set = set()

		for _ in range(100):
			val = random.randrange(max_val + 1)

			p = random.random()
			if p < add_prob:
				expected = val not in expected_set
				actual = actual_set.add(val)
				assert actual == expected, \
					f'{type_name}({expected_set})\n.add({val}):\nwant {expected}\n got {actual}'
				expected_set.add(val)
			elif p < find_prob:
				expected = val in expected_set
				actual = actual_set.contains(val)
				assert actual == expected, \
					f'{type_name}({expected_set})\n.contains({val}):\nwant {expected}\n got {actual}'
			else:
				expected = val in expected_set
				actual = actual_set.remove(val)
				assert actual == expected, \
					f'{type_name}({expected_set})\n.remove({val}):\nwant {expected}\n got {actual}'
				expected_set.discard(val)
